include/wil/cppwinrt_register_com_server.h

```diff
@@ -0,0 +1,90 @@
+//*********************************************************
+//
+//    Copyright (c) Microsoft. All rights reserved.
+//    This code is licensed under the MIT License.
+//    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+//    ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+//    TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+//    PARTICULAR PURPOSE AND NONINFRINGEMENT.
+//
+//*********************************************************
+//! @file
+//! Utilities for making managing COM server easier
+
+#ifndef __WIL_CPPWINRT_REGISTER_COM_SERVER_DEFINED
+#define __WIL_CPPWINRT_REGISTER_COM_SERVER_DEFINED
+
+#if _MSVC_LANG >= 201703L
+
+#include <winrt/base.h>
+#include <vector>
+
+namespace wil::details
+{
+template <typename T>
+struct Factory : winrt::implements<Factory<T>, IClassFactory, winrt::no_module_lock>
+{
+    HRESULT __stdcall CreateInstance(IUnknown* outer, GUID const& iid, void** result) noexcept final
+    {
+        *result = nullptr;
+
+        if (outer)
+        {
+            return CLASS_E_NOAGGREGATION;
+        }
+
+        return winrt::make_self<T>().as(iid, result);
+    }
+
+    HRESULT __stdcall LockServer(BOOL) noexcept final
+    {
+        return S_OK;
+    }
+};
+
+template <typename T = void, typename... Rest>
+void register_com_server(std::vector<DWORD>& registrations)
+{
+    auto& back = registrations.emplace_back();
+    winrt::check_hresult(
+        CoRegisterClassObject(winrt::guid_of<T>(), winrt::make<Factory<T>>().get(), CLSCTX_LOCAL_SERVER, REGCLS_MULTIPLEUSE, &back));
+    register_com_server<Rest...>(registrations);
+}
+
+template <>
+void register_com_server<void>(std::vector<DWORD>&)
+{
+}
+} // namespace wil::details
+
+namespace wil
+{
+struct com_server_revoker
+{
+    std::vector<DWORD> registrations;
+    ~com_server_revoker()
+    {
+        revoke();
+    }
+    void revoke()
+    {
+        for (auto&& registration : registrations)
+        {
+            winrt::check_hresult(CoRevokeClassObject(registration));
+        }

```

Comment:
```suggestion
        }
        registrations.clear();
```
Or make `revoke` private if nobody is supposed to call it

Reply:
In my original design it was in the dtor, but I reckon there might be corner cases where users would want to revoke earlier/manually, just for the sake of API completeness. I thought of using a bool to make `Revoke` work only once but clearing registration tokens is better.